cores/esp8266/WString.cpp

```diff
@@ -252,7 +260,22 @@ String & String::operator = (const __FlashStringHelper *pstr)
 // /*********************************************/
 
 unsigned char String::concat(const String &s) {
-    return concat(s.buffer, s.len);
+    // Special case if we're concatting ourself (s += s;) since we may end up
+    // realloc'ing the buffer and moving s.buffer in the method called
+    if (&s == this) {
+        unsigned int newlen = 2 * len;
```

Comment:
if len includes the null term, then this is 1 byte too big. If not, then this is ok.

Reply:
Len's only the actual character count.  `String` is a mishmash of Pascal and C strings, with a dash of what-the-heck-were-they-thinking added, just for fun.